### ISSUE TITLE
Lack of function to compute the "hierarchy" metric for selected nodes in structuralhole.py

### ISSUE THREAD

**Issue by sea-asteroid** [networkx/networkx#3467]
In the file networkx/networkx/algorithms/structuralholes.py, there lacks a function to return the "hierarchy" value of the nodes in the network. Hierarchy is an important metric compensating the "constraint" metric. Its definition can be founded at:

"The Network Structure of Social Capital," 2000 Research in Organizational Behavior
http://faculty.chicagobooth.edu/ronald.burt/research/files/NSSC.pdf

Hierarchy is a form of closure in which a minority of contacts, typically one or two, stand above the others for being more the source of closure.

**Comment by chenyang03** [comment]
Would it be possible to add a function to the structuralhole.py file for this? Similarily, the "efficiency" metric, i.e., the effective size divided by the number of alters in ego's network, is also useful for studying the structural holes.

**Comment by dschult** [comment]
Yes, it should be possible to add these features to the structuralhole.py file and corresponding simple tests (just to make sure it runs and works in the simplest of cases).

**Comment by mcauley-penney** [comment]
@dschult I'm currently working on an implementation for the hierarchy metric using NetworkX for work and am interested in submitting it as a PR but would be interested in a guiding force to help with it because I don't have a *very* thorough understanding of Burt's writing on it. I've referred to `Structural Holes: The Social Structure of Competition` but am primarily translating the implementation from `Java Universal Network/Graph Framework` because that is the implementation that a prior research work, which our project relies upon as a reference, used. 

If I opened a PR, would you or another author with a deeper familiarity with Burt's work be available for review?

**Comment by Shikhar-Kesharwani** [comment]
Hi @dschult and @chenyang03,

I have submitted an implementation of Burt's **Hierarchy** and **Structural Efficiency** metrics for `structuralholes` in PR #8905 (https://github.com/networkx/networkx/pull/8905).

### Summary of Changes:
- Added `hierarchy(G, nodes=None, weight=None)` implementing Burt's Coleman-Theil disorder index of local constraints.
- Added `structural_efficiency(G, nodes=None, weight=None)` implementing the ratio of effective size to degree.
- Included both a vectorized SciPy sparse matrix engine (for all-node computations) and an iterative fallback (for node subsets).
- Added comprehensive unit tests in `test_structuralholes.py` covering isolates, single contacts, star graphs, complete graphs, paths, weighted and directed networks, and sparse vs. iterative parity.

**Comment by Shikhar-Kesharwani** [comment]
Hi @dschult and @chenyang03,

I have submitted an implementation of Burt's **Hierarchy** and **Structural Efficiency** metrics for `structuralholes` in PR #8923 (https://github.com/networkx/networkx/pull/8923).

### Summary of Changes:
- Added `hierarchy(G, nodes=None, weight=None)` implementing Burt's Coleman-Theil disorder index of local constraints.
- Added `structural_efficiency(G, nodes=None, weight=None)` implementing the ratio of effective size to degree.
- Included both a vectorized SciPy sparse matrix engine (for all-node computations) and an iterative fallback (for node subsets).
- Added comprehensive unit tests in `test_structuralholes.py` covering isolates, single contacts, star graphs, complete graphs, paths, weighted and directed networks, and sparse vs. iterative parity.
